plugins/osdn/vnidallocator/allocator.go

```diff
@@ -0,0 +1,146 @@
+package vnidallocator
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/golang/glog"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/registry/service/allocator"
+
+	"github.com/openshift/openshift-sdn/plugins/osdn/util"
+)
+
+// Interface manages the allocation of VNIDs out of a range.
+// Interface should be threadsafe.
+type Interface interface {
+	Allocate(uint) error
+	AllocateNext() (uint, error)
+	Release(uint) error
+}
+
+var (
+	ErrFull            = errors.New("range is full")
+	ErrNotInRange      = errors.New("provided VNID is not in the valid range")
+	ErrAllocated       = errors.New("provided VNID is already allocated")
+	ErrMismatchedRange = errors.New("the provided VNID range does not match the current VNID range")
+)
+
+type Allocator struct {
+	vnidRange util.VNIDRange
+
+	alloc allocator.Interface
+}
+
+// Allocator implements Interface and Snapshottable
+var _ Interface = &Allocator{}
+
+// NewAllocatorCustom creates a Allocator over a VNID Range, calling allocatorFactory to construct the backing store.
+func NewAllocatorCustom(vr util.VNIDRange, allocatorFactory allocator.AllocatorFactory) *Allocator {
+	return &Allocator{
+		vnidRange: vr,
+		alloc:     allocatorFactory(int(vr.Size), vr.String()),
+	}
+}
+
+// Helper that wraps NewAllocatorCustom, for creating a range backed by an in-memory store.
+func NewInMemoryAllocator(vr util.VNIDRange) *Allocator {
+	return NewAllocatorCustom(vr, allocator.NewContiguousAllocationInterface)
+}
+
+// Free returns the count of VNID left in the range.
+func (r *Allocator) Free() int {
+	return r.alloc.Free()
+}
+
+// Allocate attempts to reserve the provided VNID. ErrNotInRange or
+// ErrAllocated will be returned if the VNID is not valid for this range
+// or has already been reserved. ErrFull will be returned if there
+// are no VNIDs left.
```

Comment:
Either (a) vnid is out of range, in which case ErrNotInRange, or (b) it's in range but already used, in which case ErrAllocated, or (c) it's in range and not already used, in which case Allocate() will succeed. It shouldn't be possible to get ErrFull.